packages/optimizer/spec/helpers/TransformerRunner.js

```diff
@@ -45,56 +46,81 @@ if (WRITE_SNAPSHOT) {
-        );
-        let expectedOutput = '';
+      // parse input and extract params
+      let input = getFileContents(join(testDir, 'input.html'));
+      if (input.startsWith(CONFIG_START_TOKEN)) {
+        const indexStartConfig = CONFIG_START_TOKEN.length;
+        const indexEndConfig = input.indexOf(CONFIG_END_TOKEN);
+        const paramsString = input.substring(indexStartConfig, indexEndConfig);
         try {
-          expectedOutput = getFileContents(expectedOutputPath);
+          params = JSON.parse(paramsString);
         } catch (e) {
-          // file doesn't exist if no snapshot has been written yet
-          // that's ok as the test will fail by comparing to an empty string
-        }
-        params = testConfig.validAmp ? {} : params;
-        params.validatorRules = await validatorRules;
-        await testConfig.transformer.transform(tree, params);
-        const actualOutput = serialize(tree, params.__format);
-        if (WRITE_SNAPSHOT) {
-          writeFileContents(expectedOutputPath, actualOutput);
-        } else {
-          expect(actualOutput).toBe(expectedOutput);
+          // no config
         }
-        if (testConfig.validAmp) {
-          const result = validatorInstance.validateString(actualOutput);
-          if (result.status !== 'PASS') {
-            fail(`Validation errors:\n\n ${JSON.stringify(result.errors, null, 2)}`);
-          }
+        // trim params from input string
+        input = input.substring(indexEndConfig + CONFIG_END_TOKEN.length);
+      }
+
+      const tree = await treeParser.parse(input);
+
+      // parse expected output
+      const expectedOutputPath = join(
+        testDir,
+        testConfig.tag ? `expected_output.${testConfig.tag}.html` : 'expected_output.html'
+      );
+      let expectedOutput = '';
+      try {
+        expectedOutput = getFileContents(expectedOutputPath);
+      } catch (e) {
+        // file doesn't exist if no snapshot has been written yet
+        // that's ok as the test will fail by comparing to an empty string
+      }
+      params = testConfig.validAmp ? {} : params;
+      params.validatorRules = await validatorRules;
+      await testConfig.transformer.transform(tree, params);
+      const actualOutput = serialize(tree, params.__format);
+      if (WRITE_SNAPSHOT) {
+        writeFileContents(expectedOutputPath, actualOutput);
+      } else {
+        expect(actualOutput).toBe(expectedOutput);
+      }
+      if (testConfig.validAmp) {
+        const result = validatorInstance.validateString(actualOutput);
+        if (result.status !== 'PASS') {
+          fail(`Validation errors:\n\n ${JSON.stringify(result.errors, null, 2)}`);
         }
-        done();
+      }
+      done();
+    };
+
+    getDirectories(testConfig.testDir).forEach((testDir) => {
+      it(basename(testDir), (done) => runTests(testDir, done));
+
+      it('works when skipNodeAndChildren returns null', function (done) {
```

Comment:
I'd prefer not to add more tests here. Instead, we could add an end-to-end test [here](https://github.com/ampproject/amp-toolbox/tree/master/packages/optimizer/spec/end-to-end) that runs a document with a template as last element through all transformers.

Reply:
sounds good